### ISSUE TITLE
Arrow function not tokenized correctly when using null in union type

### ISSUE THREAD

**Issue by simPod** [squizlabs/PHP_CodeSniffer#3316]
**Describe the bug**

phpcs shoots:

> PSR2.Methods.FunctionCallSignature.SpaceBeforeOpenBracket: Space before opening parenthesis of function call prohibited
> Squiz.Arrays.ArrayDeclaration.IndexNoNewline: Each index in a multi-line array must be on a new line

when types are declared as union types. `array|null` triggers `Squiz.Arrays.ArrayDeclaration.IndexNoNewline`. `?array` is fine.

**Code sample**
```php
<?php

declare(strict_types=1);

return [
    'this is fine' => static fn (?DateTimeImmutable $value) : ?array => $value === null
        ? null
        : [],
    'this shoots false positives' => static fn (DateTimeImmutable|null $value) : array|null => $value === null
        ? null
        : [],
];
```
**Custom ruleset**

doctrine/coding-standard v9

**To reproduce**

```
 9 | ERROR | [x] Space before opening parenthesis of function call prohibited
   |       |     (PSR2.Methods.FunctionCallSignature.SpaceBeforeOpenBracket)
 9 | ERROR | [x] Each index in a multi-line array must be on a new line (Squiz.Arrays.ArrayDeclaration.IndexNoNewline)
```

**Expected behavior**

No errors.

**Versions (please complete the following information):**
 - OS: [MacOS]
 - PHP: [8.0]
 - PHPCS: [master]
 - Standard: [Doctrine]

**Comment by gsherwood** [comment]
I think the core error causing this is the T_BITWISE_OR token not being converted to T_TYPE_UNION. That stops the arrow function being processed correctly, which causes the sniff to produce incorrect errors.

**Comment by gsherwood** [comment]
This issue was caused by the arrow function tokenizing code not allowing for the null type in a type union. It appears this was the only place that had this issue in the codebase.

Thanks a lot for the bug report. The fix will be in 3.6.1.
